libmscore/score.cpp

```diff
@@ -2067,59 +2034,132 @@ bool Score::appendScore(Score* score, bool addPageBreak, bool addSectionBreak)
-            Segment* seg = firstAppendedMeasure->getSegment(SegmentType::KeySig, tickOfAppend);
-            for (Staff* st : score->staves()) {
-                  int staffIdx = st->idx();
-                  Staff* joinedStaff = staff(staffIdx);
-                  // special case for initial "C" key signature - these have no explicit element
-                  if (!seg->element(staffIdx * VOICES)) {
-                        // no need to create new initial "C" key sig
-                        // if staff already ends in that key
-                        if (joinedStaff->key(tickOfAppend - Fraction::fromTicks(1)) == Key::C)
-                              continue;
-                        Key key = Key::C;
-                        KeySig* ks = new KeySig(this);
-                        ks->setTrack(staffIdx * VOICES);
-                        ks->setKey(key);
-                        ks->setParent(seg);
-                        addElement(ks);
-                        }
-                  // other key signatures (initial other than "C", non-initial)
-                  for (auto k : *(st->keyList())) {
-                        Fraction tick = Fraction::fromTicks(k.first);
-                        KeySigEvent key = k.second;
-                        joinedStaff->setKey(tick + tickOfAppend, key);
-                        }
+      // at first added measure, check if we need to add Clef/Key/TimeSig
+      //  this is needed if it was changed and needs to be changed back
+      int n = nstaves();
+      Fraction otick = fmb->tick(), ctick = tickOfAppend;
+      qDebug() << "Checking if something needs to be added" << otick << ctick;
+      for (int staffIdx = 0; staffIdx < n; ++staffIdx) { // iterate over all staves
+            int trackIdx = staff2track(staffIdx); // idx of irst track on the staff
+            Staff* staff = this->staff(staffIdx);
+            Staff* ostaff = score->staff(staffIdx);
+            
+            // check if key signature needs to be changed
+            if (!(ostaff->key(otick) == staff->key(ctick))) {
+
+                  Segment* ns = firstAppendedMeasure->undoGetSegment(SegmentType::KeySig, ctick);
+                  KeySigEvent nkse = KeySigEvent(ostaff->keySigEvent(otick));
+                  KeySig* nks = new KeySig(this);
+                  nks->setScore(this);
+                  nks->setTrack(trackIdx);
+
+                  nks->setKeySigEvent(nkse);
+                  staff->setKey(ctick, nkse);
+                  ns->add(nks);
+                  //undoAddElement(nks);
+                  }
+            // check if a key signature is present but is spurious (i.e. no actual change)
+            else if (staff->currentKeyTick(ctick) == ctick &&
+                     staff->key(ctick - Fraction::fromTicks(1)) == ostaff->key(otick)) {
+                        Segment* ns = firstAppendedMeasure->first(SegmentType::KeySig);
+                        if (ns) ns->remove(ns->element(trackIdx));
```

Comment:
new line for body

Reply:
@JoshuaBonn1 I do not understand what you mean here.